### ISSUE TITLE
[Issue]: Core dumped in fresh installation when verifying torch settings for RX 7900 XTX

### ISSUE THREAD

**Issue by evshiron** [vladmandic/automatic#1929]
### Issue Description

```bash
git clone https://github.com/vladmandic/automatic
cd automatic
# reset to the commit (master at the time)
git reset --hard fc3d96f6f6a0007edc5f1531f1244cf5034bf9da

# install torch nightly by installing torchvision nightly, which depends on torch nightly from yesterday
cat << __EOF > change1.diff
diff --git a/installer.py b/installer.py
index d235bcc6..280130e6 100644
--- a/installer.py
+++ b/installer.py
@@ -315,7 +315,7 @@ def check_torch():
         log.info('AMD ROCm toolkit detected')
         os.environ.setdefault('PYTORCH_HIP_ALLOC_CONF', 'garbage_collection_threshold:0.8,max_split_size_mb:512')
         os.environ.setdefault('TENSORFLOW_PACKAGE', 'tensorflow-rocm')
-        torch_command = os.environ.get('TORCH_COMMAND', 'torch==2.0.1 torchvision==0.15.2 --index-url https://download.pytorch.org/whl/rocm5.4.2')
+        torch_command = os.environ.get('TORCH_COMMAND', 'torchvision --pre --index-url https://download.pytorch.org/whl/nightly/rocm5.6')
         xformers_package = os.environ.get('XFORMERS_PACKAGE', 'none')
     elif allow_ipex and (args.use_ipex or shutil.which('sycl-ls') is not None or os.environ.get('ONEAPI_ROOT') is not None or os.path.exists('/opt/intel/oneapi')):
         args.use_ipex = True # pylint: disable=attribute-defined-outside-init
__EOF

git apply change1.diff

HIP_VISIBLE_DEVICES=0 ./webui.sh --debug --listen
# y to download the default model

# DEBUG    Verifying Torch settings                                                                                                                        
# Segmentation fault (core dumped)
```

It failed during `test_fp16()` in `modules/devices.py`.

However, if I add one line to `webui.py`, `test_fp16()` will pass and the WebUI works perfectly.

The `HIP_VISIBLE_DEVICES=0` is needed because I have an Vega iGPU which doesn't work with AI.

```bash
cat << __EOF > change2.diff
diff --git a/webui.py b/webui.py
index 72e878c4..1e20a275 100644
--- a/webui.py
+++ b/webui.py
@@ -16,6 +16,7 @@ local_url = None
 
 errors.log.debug('Loading Torch')
 import torch # pylint: disable=C0411
+errors.log.debug(torch.randn(2, 2).to(0))
 try:
     import intel_extension_for_pytorch as ipex # pylint: disable=import-error, unused-import
 except Exception:
__EOF

git apply change2.diff

HIP_VISIBLE_DEVICES=0 ./webui.sh --debug --listen
```

Do you have an idea about what has happened in this scenario?

### Version Platform Description

Ubuntu 22.04 with ROCm 5.6 installed.

Previous installation of SD:Next works well, even if I use the same commit and the same versions of `torch` and `torchvision`.

```bash
(venv) $ pip3 freeze | grep torch
open-clip-torch==2.20.0
pytorch-lightning==1.9.4
torch==2.1.0.dev20230802+rocm5.6
torchdiffeq==0.2.3
torchmetrics==1.0.2
torchsde==0.2.5
torchvision==0.16.0.dev20230803+rocm5.6
```

### Acknowledgements

- [X] I have read the above and searched for existing issues
- [X] I confirm that this is classified correctly and its not an extension or diffusers-specific issue

**Comment by vladmandic** [comment]
first, why modify installer to specify torch when you do exactly the same by setting an environment variable?
e.g. `export TORCH_COMMAND='torchvision --pre --index-url https://download.pytorch.org/whl/nightly/rocm5.6'`

and you're saying that `test_fp16` fail **unless** you add `torch.randn(2, 2).to(0)`?
i'll take a look, but that makes no sense.
btw, where did you get the idea?

**Comment by evshiron** [comment]
I was making a fork trying to detect environment variables for ROCm automatically so I edited the code by myself, and this issue occurred when I started testing.

I had no idea why it happened too so I did copy-and-paste with a testing snippet to check if it would work somewhere in the code.

`torch.randn(2, 2).to(0)` can be any other `torch` GPU op, for example swapping it with lines from `test_fp16` works too. But if we don't call any of them here, `test_fp16()` will fail later on my end.

**Comment by evshiron** [comment]
Oh, I move the testing lines after `import pytorch_lightning` just now and it starts to fail.

**Comment by vladmandic** [comment]
that makes even less sense as test do not even use torch lightning

**Comment by evshiron** [comment]
```
$ HIP_VISIBLE_DEVICES=0 python3
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import pytorch_lightning as pl
2023-08-04 15:18:26.475322: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-08-04 15:18:26.500181: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
>>> torch.randn(2, 2).to(0)
Segmentation fault (core dumped)
```

```
HIP_VISIBLE_DEVICES=0 python3
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.randn(2, 2).to(0)
tensor([[1.4502, 1.3093],
        [0.5519, 1.3963]], device='cuda:0')
>>> import pytorch_lightning as pl
2023-08-04 15:19:19.548577: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-08-04 15:19:19.574138: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
>>> torch.randn(2, 2).to(0)
tensor([[ 0.8422,  0.7251],
        [ 0.3764, -1.4850]], device='cuda:0')
```

Maybe it's an issue with `tensorflow-rocm`, which we added Navi 31 support months ago but haven't got shipped in `tensorflow-rocm`. I will do a fresh installation with tensorflow for CPU and report back later.

**Comment by vladmandic** [comment]
looks like you're on a right track, tensorflow cpu seems to work.

**Comment by evshiron** [comment]
Yeah. I was using `tensorflow` without `-rocm`, which is said to work with both CPU and GPU (CUDA of course) according to https://www.tensorflow.org/install/pip.

Maybe the current version of `tensorflow-rocm` set up bad HSA in underlying HIP libs during import and thus affects `torch` functionality.

```
$ HIP_VISIBLE_DEVICES=0 python3
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import tensorflow
2023-08-04 17:25:06.588772: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-08-04 17:25:06.612739: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
>>> torch.randn(2, 2).to(0)
Segmentation fault (core dumped)
```

**Comment by vladmandic** [comment]
quite possibly.
not sure how to proceed here as there is not much i can do about that?

**Comment by vladmandic** [comment]
I'm not against option 2 - if there is note in the code why it's there and there is a followup few months later to see if it's still necessary or it can be removed.

On the other hand, not sure on actual benefits of tensorflow-rocm overall?

**Comment by evshiron** [comment]
`tensorflow-rocm` is broken on Navi 31 at the moment and I actually haven't explicitly used it in Automatic.

I am not sure if training Lora/Lyco uses TensorFlow but the default process works with `tensorflow` without `-rocm` on my end.

Users of CDNA and RDNA 2 might benefit from it though.

**Comment by vladmandic** [comment]
No, lora/lyco training does not use tensorflow at all (it does use tf board for loss logging)

Tf is only used in some auxiliary workflows, background removal, some (not all) controlnet models, etc.

**Comment by vladmandic** [comment]
any updates here?

**Comment by evshiron** [comment]
Sorry for being late. I tried:

* Common ControlNets
* LoRA/LyCO training
* Rembg
* Diffusers and SD-XL

All of them worked pretty well, so there was nothing special to report.

**Comment by vladmandic** [comment]
that's good.
so how do you want to proceed in this issue?

**Comment by evshiron** [comment]
I prefer option 2.

Because it works for both us Navi 3x users and those supported by `tensorflow-rocm`, and should not have side effects to users of other backends. It just looks a bit tricky.

Btw, I discover that the Diffusers version required to do training is rolled back further, would you mind giving a quick explanation here or pointing me to associated issues if any?

**Comment by vladmandic** [comment]
i've added torch init as suggested.
re: diffusers version and training - training is basically kohya_ss and it relies on old format of state_dict that was changed in newer versions of diffusers. kohya promised he'll update his code, still waiting.

**Comment by evshiron** [comment]
@vladmandic 

Thanks for the info. The `requirements.txt` do update in [his dev branch](https://github.com/kohya-ss/sd-scripts/blob/dev/requirements.txt ), and it looks like it will be merged soon.

**Comment by evshiron** [comment]
@vladmandic 

I try integrating the dev branch of kohya's scripts, and with https://github.com/are-we-gfx1100-yet/automatic/commit/94ab354dbabfd4e31fb4efb88acdcc8e77f79f8d, LoRA and LyCORIS training should work. In case you need it.
